File: golang-devildyw/middle/L48旋转图像.go
package main

func main() {
	rotate([][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}})
}

func rotate(matrix [][]int) {
	// 感觉这是一道极其吃数学思维的题 能想出先转置在水平翻转
	n := len(matrix)

	// 先转置
	for i := 0; i < n; i++ {
		for j := 0; j <= i; j++ {
			matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
		}
	}
	// 再水平翻转 逐行水平翻转
	for i := 0; i < n; i++ {
		swap(matrix[i])
	}
}

func swap(row []int) {
	i, j := 0, len(row)-1
	for i < j {
		row[i], row[j] = row[j], row[i]
		i++
		j--
	}
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L48旋转图像.java
package top.dyw.leetcode.Hot100;

public class L48旋转图像 {
    public void rotate(int[][] matrix) {
        if (matrix.length == 0) {
            return;
        }
        // 对角线对称再水平对称
        int n = matrix.length;
        // 对角线对称
        for (int i=0; i<n; i++) {
            for (int j=0; j<=i; j++) {
                int temp = matrix[j][i];
                matrix[j][i] = matrix[i][j];
                matrix[i][j] = temp;
            }
        }

        // 再左右交换
        for (int i=0; i<n; i++) {
            swap(matrix[i]);
        }
    }

    public void swap(int[] nums) {
        for (int i=0, j=nums.length-1; i<j; i++,j--) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }
}
